### PR TITLE
Fused MHA Support in XLA GPU: Stream Executor changes

### PR DESCRIPTION
**Pull request by AyanmoI**
cuDNN currently supports the following patterns for Multi-headed attention:

BMM1 - BMM2
BMM1 - Scale - Bias - Mask - Softmax - BMM2
BMM1 - Scale - Bias - Mask - Softmax - Dropout - BMM2
BMM1 - Scale - Mask - Softmax - BMM2
BMM1 - Scale - Mask - Softmax - Dropout - BMM2
BMM1 - Softmax - Dropout - BMM2

This PR adds support for the stream executor for these patterns.

### CODE REVIEW

**Review by cheshire — APPROVED** [review]

**Review by cheshire — APPROVED** [review]

**Comment by cheshire** [comment]
I think @akuegel is also fixing this internally.

**Comment by akuegel** [comment]
> I think @akuegel is also fixing this internally.

I just checked, and there are differences in the fixes. Let's import again and see whether additional fixes will be needed, and if not, we can take it as is.

File: tensorflow/compiler/xla/stream_executor/dnn.cc [hunk 1]
```diff
@@ -482,6 +542,169 @@ std::string AlgorithmConfig::ToString() const {
...
+    minor_to_major_vec[i] = minor_to_major[i];
+  }
+  return TensorDescriptor(type, dims, minor_to_major_vec);
+}
+
+const std::string TensorDescriptor::ToString() const {
+  return absl::StrFormat("{dimensions: %s minor_to_major: %s}",
+                         absl::StrJoin(dimensions(), ","),
+                         absl::StrJoin(minor_to_major(), ","));
+};
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.cc:608]: FYI: our linter is complaining about the extra ';' here, I will remove it in the internally adjusted version of this PR.

File: tensorflow/compiler/xla/stream_executor/dnn.cc [hunk 2]
```diff
@@ -482,6 +542,169 @@ std::string AlgorithmConfig::ToString() const {
...
+      TensorDescriptor::For(type, dimensions, minor_to_major), batch_dims_vec,
+      contracting_dims_vec);
+}
+
+const std::string MatmulTensorDescriptor::ToString() const {
+  return absl::StrFormat(
+      "{%s, batch_dimension_numbers: %s contracting_dim: %s}",
+      tensor_.ToString(), absl::StrJoin(batch_dimension_numbers_, ","),
+      absl::StrJoin(contracting_dim_, ","));
+};
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.cc:706]: FYI: our linter is complaining about the extra ';' here, I will remove it in the internally adjusted version of this PR.

File: tensorflow/tsl/protobuf/dnn.proto [hunk 3]
```diff
@@ -174,3 +174,9 @@ message ConvolutionDescriptorProto {
   // Tensorflow node name, same as in NodeDef, for debugging purposes.
   string name = 7;
 }
+
+// FusedMHAKind kind
+enum FusedMHAKind {
+  BMM1_OUTPUT_INPUT_TYPE = 0;
```
Comment (akuegel) [inline tensorflow/tsl/protobuf/dnn.proto:180]: Best practice is to add a default unknown entry, so this will become:

  BMM1_OUTPUT_UNKNOWN = 0;
  BMM1_OUTPUT_INPUT_TYPE = 1;
  BMM1_OUTPUT_FLOAT = 2;

**Comment by akuegel** [comment]
This has been merged. I had to also apply some fixes suggested by ClangTidy as it was also blocking the submit.

**Comment by AyanmoI** [comment]
> pushed a commit to google/tsl that referenced this pull request

@akuegel what are the issues? We can work on fixing them and re-merge.

**Comment by AyanmoI** [comment]
Can we re-open this PR to work on fixing the issues? I don't see a way to re-open this.

**Comment by hawkinsp** [comment]
It fails to build under CUDA 11.4 /CuDNN 8.2:

```
external/xla/xla/stream_executor/cuda/cuda_dnn.cc: At global scope:
[external/xla/xla/stream_executor/cuda/cuda_dnn.cc:3500](https://cs.corp.google.com/piper///depot/google3/external/xla/xla/stream_executor/cuda/cuda_dnn.cc?l=3500):5: error: 'cudnnBackendTensorReordering_t' has not been declared
 3500 |     cudnnBackendTensorReordering_t cudnn_tensor_order_type =
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[external/xla/xla/stream_executor/cuda/cuda_dnn.cc:3501](https://cs.corp.google.com/piper///depot/google3/external/xla/xla/stream_executor/cuda/cuda_dnn.cc?l=3501):9: error: 'CUDNN_TENSOR_REORDERING_NONE' was not declared in this scope; did you mean 'cudnn_frontend::cudnnBackendTensorReordering_t::CUDNN_TENSOR_REORDERING_NONE'?
 3501 |         CUDNN_TENSOR_REORDERING_NONE,
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         cudnn_frontend::cudnnBackendTensorReordering_t::CUDNN_TENSOR_REORDERING_NONE
In file included from [bazel-out/k8-opt/bin/external/cudnn_frontend_archive/_virtual_includes/cudnn_frontend/third_party/cudnn_frontend/include/cudnn_frontend_ConvDesc.h:35](https://cs.corp.google.com/piper///depot/google3/bazel-out/k8-opt/bin/external/cudnn_frontend_archive/_virtual_includes/cudnn_frontend/third_party/cudnn_frontend/include/cudnn_frontend_ConvDesc.h?l=35),
                 from [bazel-out/k8-opt/bin/external/cudnn_frontend_archive/_virtual_includes/cudnn_frontend/third_party/cudnn_frontend/include/cudnn_frontend.h:100](https://cs.corp.google.com/piper///depot/google3/bazel-out/k8-opt/bin/external/cudnn_frontend_archive/_virtual_includes/cudnn_frontend/third_party/cudnn_frontend/include/cudnn_frontend.h?l=100),
                 from [external/xla/xla/stream_executor/cuda/cuda_dnn.cc:56](https://cs.corp.google.com/piper///depot/google3/external/xla/xla/stream_executor/cuda/cuda_dnn.cc?l=56):
[bazel-out/k8-opt/bin/external/cudnn_frontend_archive/_virtual_includes/cudnn_frontend/third_party/cudnn_frontend/include/cudnn_frontend_utils.h:376](https://cs.corp.google.com/piper///depot/google3/bazel-out/k8-opt/bin/external/cudnn_frontend_archive/_virtual_includes/cudnn_frontend/third_party/cudnn_frontend/include/cudnn_frontend_utils.h?l=376):5: note: 'cudnn_frontend::cudnnBackendTensorReordering_t::CUDNN_TENSOR_REORDERING_NONE' declared here
  376 |     CUDNN_TENSOR_REORDERING_NONE,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~

etc.
```

**Comment by AyanmoI** [comment]
@akuegel After reopening the PR, why does it now have only one file? Was this partially merged?

**Comment by AyanmoI** [comment]
It looks like only the last commit got reverted. I don't think the issue (of build failing with CUDA 11.4 and CuDNN 8.2) will have resolved with reverting the last commit only.

**Comment by akuegel** [comment]
> 

Actually all of it was reverted:

https://github.com/tensorflow/tensorflow/commit/7a4d2e8b1a503a45ff86d40b0387fc832302379f

Why it doesn't show properly merged in this PR, I don't know :-(
Maybe it is easier to create a new PR?

**Comment by AyanmoI** [comment]
> > 
> 
> Actually all of it was reverted:
> 
> [7a4d2e8](https://github.com/tensorflow/tensorflow/commit/7a4d2e8b1a503a45ff86d40b0387fc832302379f)
> 
> Why it doesn't show properly merged in this PR, I don't know :-( Maybe it is easier to create a new PR?

Yeah..I guess. Let me do that.

**Comment by AyanmoI** [comment]
@akuegel @cheshire opened a new PR https://github.com/tensorflow/tensorflow/pull/60139
